File: NumGuess.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class NumGuess {
	static String name;
	static int limit;
	static int number;
	static int tries;
	static int max_tries;

	public static String readLine() {
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		try {
			return reader.readLine();
		} catch (IOException e) {
			return "";
		}
	}

	public static void main(String[] args) {
		System.out.printf("Welcome to NumGuess Java terminal version!\n\n");
		System.out.printf("Enter your name: ");

		name = readLine();
		if (name.isEmpty()) name = "Player";

		System.out.printf("\nWelcome %s, enter upper limit: ", name);

		try {
			limit = Integer.parseInt(readLine(), 10);
			if (limit < 10) limit = 10;
		} catch (NumberFormatException e) {
			limit = 10;
		}

		max_tries = (int) Math.floor(Math.log(limit) / Math.log(2)) + 1;

		while (true) {
			tries = 0;
			number = (int) Math.floor(Math.random() * limit) + 1;
			System.out.printf("\nGuess my number between 1 and %d!\n", limit);

			while (true) {
				int guess;
				System.out.printf("\nGuess: ");

				try {
					guess = Integer.parseInt(readLine(), 10);

					if (guess < 1 || guess > limit) {
						System.out.printf("Out of range.");
					} else {
						tries++;
						if (guess < number) {
							System.out.printf("Too low!");
						} else if (guess > number) {
							System.out.printf("Too high!");
						} else break;
					}
				} catch (NumberFormatException e) {
					System.out.printf("That's just plain wrong.");
				}
			}

			System.out.printf("\nWell done %s, you guessed my number from %d %s!\n", name, tries, tries == 1 ? "try" : "tries");

			if (tries == 1) {
				System.out.printf("You're one lucky bastard!");
			} else if (tries < max_tries) {
				System.out.printf("You are the master of this game!");
			} else if (tries == max_tries) {
				System.out.printf("You are a machine!");
			} else if (tries <= max_tries * 1.1f) {
				System.out.printf("Very good result!");
			} else if (tries > limit) {
				System.out.printf("I find your lack of skill disturbing!");
			} else {
				System.out.printf("Try harder, you can do better!");
			}

			System.out.printf("\nPlay again [y/N]? ");

			if (!readLine().toUpperCase().equals("Y")) break;
		}

		System.out.printf("\nOkay, bye.\n");
	}
}

File: numguess.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"time"
	"math"
	"math/rand"
)

func main() {
	var err error
	rand.Seed(time.Now().Unix())
	fmt.Printf("Welcome to NumGuess Go version!\n\n")

	print("Enter your name: ")
	name := readLine()
	if name == "" { name = "Player" }

	fmt.Printf("\nWelcome %v, enter upper limit: ", name)
	limit, _ := readInt(10)
	if limit < 10 { limit = 10 }
	maxTries := int(math.Log2(float64(limit))) + 1

	for {
		fmt.Printf("\nGuess my number between 1 and %v!\n\n", limit);
		tries, num := 0, rand.Intn(limit) + 1
		guess := 0

		for guess != num {
			fmt.Printf("Guess: ")
			guess, err = readInt(0)

			if err != nil {
				fmt.Printf("That's just plain wrong.\n")
			} else {
				if guess < 1 || guess > limit {
					fmt.Printf("Out of range.\n")
				} else {
					tries++
					switch {
						case guess < num: fmt.Printf("Too low!\n")
						case guess > num: fmt.Printf("Too high!\n")
					}
				}
			}
		}

		triesWord := "try"
		if tries != 1 { triesWord = "tries" }
		fmt.Printf("\nWell done %v, you guessed my number from %v %v!\n", name, tries, triesWord)

		switch {
			case tries == 1:
				fmt.Printf("You're one lucky bastard!")
			case tries < maxTries:
				fmt.Printf("You are the master of this game!")
			case tries == maxTries:
				fmt.Printf("You are a machine!")
			case tries <= int(float64(maxTries) * 1.1):
				fmt.Printf("Very good result!")
			case tries > limit:
				fmt.Printf("I find your lack of skill disturbing!")
			default:
				fmt.Printf("Try harder, you can do better!")
		}

		fmt.Printf("\nPlay again [y/N]? ")
		if strings.ToUpper(readLine()) != "Y" { break }
	}

	fmt.Printf("\nOkay, bye.")
}

// Reads a whole line from stdin
func readLine() string {
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	return strings.Trim(line, "\n\r")
}

func readInt(def int) (int, error) {
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	_, err := fmt.Sscan(strings.Trim(line, "\n\r"), &def)
	return def, err
}
